l1-contracts/src/core/libraries/staking/StakingLib.sol

```diff
@@ -190,8 +196,32 @@ library StakingLib {
     uint256 amount = store.gse.DEPOSIT_AMOUNT();
 
     store.stakingAsset.transferFrom(msg.sender, address(this), amount);
-    store.stakingAsset.approve(address(store.gse), amount);
-    store.gse.deposit(_attester, _withdrawer, _onCanonical);
+    store.entryQueue.enqueue(_attester, _withdrawer, _onCanonical);
+    emit IStakingCore.ValidatorQueued(_attester, _withdrawer);
+  }
+
+  function flushEntryQueue(uint256 _maxAddableValidators) internal {
+    Epoch currentEpoch = TimeLib.epochFromTimestamp(Timestamp.wrap(block.timestamp));
+    StakingStorage storage store = getStorage();
+    require(
+      store.nextFlushableEpoch <= currentEpoch, Errors.Staking__QueueAlreadyFlushed(currentEpoch)
+    );
+    store.nextFlushableEpoch = currentEpoch + Epoch.wrap(1);
+    uint256 amount = store.gse.DEPOSIT_AMOUNT();
+
+    uint256 queueLength = store.entryQueue.length();
+    uint256 numToDequeue = _maxAddableValidators > queueLength ? queueLength : _maxAddableValidators;
+    store.stakingAsset.approve(address(store.gse), amount * numToDequeue);
+    for (uint256 i = 0; i < numToDequeue; i++) {
+      DepositArgs memory args = store.entryQueue.dequeue();
+      try store.gse.deposit(args.attester, args.withdrawer, args.onCanonical) {
+        emit IStakingCore.Deposit(args.attester, args.withdrawer, amount);
+      } catch {
+        store.stakingAsset.transfer(args.withdrawer, amount);
+        emit IStakingCore.FailedDeposit(args.attester, args.withdrawer);
+      }
+    }
+    store.stakingAsset.approve(address(store.gse), 0);
```

Comment:
In practice, we could probably get away with just doing an infinite approval at the constructor, just something not sitting with me right when doing that 😆

Reply:
on this note, there is a case where a malicious actor can craft the deposit call such that it will run out of gas, and the whole call will succeed, removing them from the queue

Reply:
it is also possible in the testnet implementation but there is no stake, and people can re-add themselves to the queue so i its not a risk, it is a bit more here